### ISSUE TITLE
OIDC cache returns old token even after the expiration of refresh token in k8s go-client oidc auth provider

### ISSUE THREAD

**Issue by DheerajSShetty** [kubernetes/kubernetes#78494]
**What happened**:
When a client's oidc token has expired, even on passing the latest token to create a client, newOIDCAuthProvider() returns the old client with the expired tokens.

In https://github.com/kubernetes/client-go/blob/79226fe1949a01066ee9e3a3f4c53546d72e1194/plugin/pkg/client/auth/oidc/oidc.go#L122
newOIDCAuthProvider()

Cached Client is retrieved as shown below

// Check cache for existing provider.
if provider, ok := cache.getClient(issuer, clientID); ok {
return provider, nil
}

However, when after token expiration, when we make a new call to create a client, the refresh token would be different. Since we use (key := cacheKey{issuer, clientID}) as cached key in this case it will still return old token.

**What you expected to happen**:
On passing the new refresh token, the code should return new client

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a go-client to a cluster with oidc based auth
2. Use the client for cluster based operations, this verifies client works.
3. Do not stop the process using the client.
4. After the token expiry, the code will use the new Kubeconfig with the new token to create a new client.
5. Use the client for cluster based operation. It will fail with token expired error.

**Anything else we need to know?**:
Similar issue opened few years back.
https://github.com/kubernetes/client-go/issues/268

**Environment**:
- Go client version.
    `@k8s.io/client-go` v11.0.0
- Kubernetes version (use `kubectl version`):
  1.14.1
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
  Ubuntu
- Kernel (e.g. `uname -a`):
 4.15.0-1037-aws
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by DheerajSShetty** [comment]
@liggitt

**Comment by hainesc** [comment]
Client in OIDC is not the end-user, it is the third party when the end-user take authentication(Authn) from the oidc service provider. There is no problem with the cache here. Yeah, OIDC is somewhat complex, there is no need to refresh client, by normal. Refresh the token, it is the key that you access the client.

**Comment by liggitt** [comment]
> There is no problem with the cache here. Yeah, OIDC is somewhat complex, there is no need to refresh client, by normal. Refresh the token, it is the key that you access the client.

the token is not included in the cache key, and the cached provider has the stale config with the old token

**Comment by hainesc** [comment]
It refreshes the id token here.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go#L268

The provider cache the id_token, but when the function `idToken()` called, it checks `idTokenExpired ` first, if valid, return the cache one, else refresh the ID token if refresh token exists.

**Comment by liggitt** [comment]
right, and when the refresh token expires, the provider is dead in the water, and calling newOIDCAuthProvider() even with a new config with a new refresh token returns the dead cached provider

**Comment by hainesc** [comment]
> right, and when the refresh token expires, the provider is dead in the water, and calling newOIDCAuthProvider() even with a new config with a new refresh token returns the dead cached provider

Right, I think we should add this piece of code at https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go#L120
```go
	idToken := cfg[cfgIDToken]
	if idToken != "" {
		// update the cache
	}
        refreshToken := cfg[cfgRefreshToken]
	if refreshToken != "" {
		// update the cache
	}

**Comment by DheerajSShetty** [comment]
if we pass idToken and refreshToken, we will keep updating the cache everytime. This solution works though.

**Comment by hainesc** [comment]
I will have a pr soon, thanks!

**Comment by YanzhaoLi** [comment]
Any updates on this issue?

**Comment by YanzhaoLi** [comment]
Why do we need a cache here, why not just use a new client each time? Are there any drawbacks?

**Comment by jack-evans** [comment]
Are there any updates on this issue? I have found I am hitting this quite consistently where my refresh token expires and the only way to fix it is to restart my applications

**Comment by enj** [comment]
/assign @enj

To clarify next steps on getting this fixed.

**Comment by s-urbaniak** [comment]
To sum up the current state of this issue and my understanding. Generally, the current code is able to self-maintain tokens:

1. ID tokens: renew using the cached refresh token upon its expiry: https://github.com/kubernetes/client-go/blob/75fea27a27a0fb881c29cedc64d623084b124bc2/plugin/pkg/client/auth/oidc/oidc.go#L225-L234
2. refresh tokens: renew the refresh token from the ID token response from step 1.: https://github.com/kubernetes/client-go/blob/75fea27a27a0fb881c29cedc64d623084b124bc2/plugin/pkg/client/auth/oidc/oidc.go#L275-L279

This assumes invocations and client roundtrips happen frequently enough such that the refresh token stays valid and does not expire. However, if the refresh token expires before ID token renewal or if the refresh token is invalidated externally, step 1. fails in the first place.

Instanciation of a new client in this case doesn't help because the client-specific provider is being reused which caches the invalid refresh token: https://github.com/kubernetes/client-go/blob/dae61bee3a0a9209ceca237eebed9eaeabc40240/plugin/pkg/client/auth/oidc/oidc.go#L123.

https://github.com/kubernetes/kubernetes/pull/78509 suggests to refresh the provider's internal cache when instanciating a new client by updating the existing provider cache with the new token material (both refresh and id token).

This approach sounds still sounds reasonable to as the provider's internal cache is already synchronized, protected by a mutex, and updated anyways during roundtrips.

Discussing OOB with @enj suggested to use a file lock mechanism but the internal cache is already protected in-process.

Suggestion from my side: I would like to revive https://github.com/kubernetes/kubernetes/pull/78509 and submit a very similar in-process update mechanism, wdyt @enj?

**Comment by djain13** [comment]
@s-urbaniak @enj 
My team at IBM is also hitting into this issue. Is someone working on the fix? 

Cheers
Dharmesh

**Comment by slaskawi** [comment]
I had a discussion with @s-urbaniak and I'd like to volunteer for this issue. 

After carefully evaluating the previous attempt to fix it (https://github.com/kubernetes/kubernetes/issues/78494), I took @liggitt 's suggestion into consideration and proposed a slightly different approach: https://github.com/kubernetes/kubernetes/pull/108794

Please have a look at it when you find a free slot.

**Comment by lipmanp** [comment]
is it good to let oidc client cache auto remove the client if the token expired?
